pkg/gui/controllers/scroll_off_margin.go

```diff
@@ -0,0 +1,40 @@
+package controllers
+
+import (
+	"github.com/jesseduffield/lazygit/pkg/gui/types"
+)
+
+// To be called after pressing up-arrow; checks whether the cursor entered the
+// top scroll-off margin, and so the view needs to be scrolled up one line
+func checkScrollUp(view types.IViewTrait, scrollOffMargin int, lineIdxBefore int, lineIdxAfter int) {
+	if !insideTopMargin(view, scrollOffMargin, lineIdxBefore) &&
+		insideTopMargin(view, scrollOffMargin, lineIdxAfter) {
+		view.ScrollUp(1)
+	}
+}
+
+// To be called after pressing down-arrow; checks whether the cursor entered the
+// bottom scroll-off margin, and so the view needs to be scrolled down one line
+func checkScrollDown(view types.IViewTrait, scrollOffMargin int, lineIdxBefore int, lineIdxAfter int) {
+	if !insideBottomMargin(view, scrollOffMargin, lineIdxBefore) &&
+		insideBottomMargin(view, scrollOffMargin, lineIdxAfter) {
+		view.ScrollDown(1)
+	}
+}
+
+func insideTopMargin(view types.IViewTrait, scrollOffMargin int, lineIdx int) bool {
+	startIdx, length := view.ViewPortYBounds()
+	if length < scrollOffMargin*2 {
```

Comment:
Why `*2`?

Reply:
This is meant to disable the feature if the view is so small that it doesn't have enough room for both the top and bottom margins.

Reply:
I added another fixup (8bcc6da31a) which caps the margin at half the view height instead of disabling the feature when the view is not tall enough. This allows people to set the config to a very large value and effectively center the cursor in the middle, like you [can in vim](https://news.ycombinator.com/item?id=12647609).